packages/plugin-workspace-tools/sources/commands/foreach.ts

```diff
@@ -126,14 +132,27 @@ export default class WorkspacesForeachCommand extends BaseCommand {
       ? project.topLevelWorkspace
       : cwdWorkspace!;
 
+    if (configuration.projectCwd === null)
+      throw new UsageError(`This command can only be run from within a Yarn project`);
+
+    const root = await gitUtils.fetchRoot(configuration.projectCwd);
+    const base = root !== null
+      ? await gitUtils.fetchBase(root, {baseRefs: [typeof this.since === `string` ? this.since : await gitUtils.fetchDefaultBranch(root)]})
+      : null;
+
+    const prospectiveWorkspaces = this.since && root !== null
+      ? Array.from(await gitUtils.fetchChangedWorkspaces(root, {base: base!.hash, project}))
+      : [rootWorkspace];
+
     const fromPredicate = (workspace: Workspace) => micromatch.isMatch(structUtils.stringifyIdent(workspace.locator), this.from);
     const fromCandidates: Array<Workspace> = this.from.length > 0
-      ? [rootWorkspace, ...rootWorkspace.getRecursiveWorkspaceChildren()].filter(fromPredicate)
-      : [rootWorkspace];
+      ? prospectiveWorkspaces.filter(fromPredicate)
```

Comment:
I still need to find a clean way to handle the case where `--from` is set but `--since` is not.

Reply:
Done.